packages/did-mailto/src/index.js

```diff
@@ -0,0 +1,70 @@
+export const foo = 1
+
+/**
+ * create a did:mailto from an email address
+ *
+ * @param {import("./types").EmailAddress} email
+ * @returns {import("./types").DidMailto}
+ */
+export function fromEmail(email) {
+  const { domain, local } = parseEmail(email)
+  const did = /** @type {const} */ (
+    `did:mailto:${encodeURIComponent(domain)}:${encodeURIComponent(local)}`
+  )
+  return did
+}
+
+/**
+ * @param {string} email
+ */
+function parseEmail(email) {
+  const atParts = email.split('@')
+  if (atParts.length < 2) {
+    throw new TypeError(
+      `expected at least 2 @-delimtied segments, but got ${atParts.length}`
+    )
+  }
+  const domain = atParts.at(-1) ?? ''
+  const local = atParts.slice(0, -1).join('@')
+  return { domain, local }
+}
+
+/**
+ * @param {import("./types").DidMailto} did
+ * @returns {import("./types").EmailAddress}
+ */
+export function toEmail(did) {
+  const parts = did.split(':')
+  if (parts[1] !== 'mailto') {
+    throw new Error(`DID ${did} is not a mailto did.`)
+  }
+  return `${decodeURIComponent(parts[3])}@${decodeURIComponent(parts[2])}`
+}
+
+/**
```

Comment:
what's this `email()` function for? it seems like it pulls the email address apart just to put it back together again. 

If we get rid of this then we can also get rid of https://github.com/web3-storage/w3up/pull/722/files#diff-7136e328e23c0e6a0f91abc8e3007589dc0774d34a249b2001f9126d7d364029 which would be great. 

At the very least, could you add a comment here explaining why this exists?

Reply:
it accepts a string and returns an `EmailAddress` (which is more specific than a string)

Reply:
added a comment. https://github.com/web3-storage/w3up/pull/722/commits/3b1ff2b0ec732916bbf421ccde3bddf202109338#diff-6caa234bd554caeddcc0602f2a8ffc66e596debc4b1372f0fdfdb067b3d9f3b4R30

Reply:
ah I see. it feels kind of weird to be doing runtime work (ie, pulling the string apart and then reassembling it into the same string) for what should be compile-time typechecking (ie, ensuring that strings that are emails have the EmailAddress type during compilation) 

I could see this being valuable if we thought we might change the EmailAddress type in the future to be more complicated without changing all these APIs - is that the kind of flexibility you're hoping to buy with this?

If that's not an explicit goal, I wonder if it might be better to have `fromEmail` take a `string | EmailAddress` and throw if the `string` isn't an actual `EmailAddress`?

I guess one nice thing about factoring things this way is that you make that string -> EmailAddress conversion optional, only needed by folks who haven't already checked and case the raw string. This line of thinking makes me think that maybe the name is the problem here - `email()` just feels like a puzzling thing to have to call on my `string` before I can use it with this API - maybe `stringToEmailAddress` or `ensureEmailAddress` or something would read a bit better?

I'm also fine with just leaving as-is, but I may propose a name change in the future :)

Reply:
> I could see this being valuable if we thought we might change the EmailAddress type in the future to be more complicated without changing all these APIs - is that the kind of flexibility you're hoping to buy with this?

yeah

> I guess one nice thing about factoring things this way is that you make that string -> EmailAddress conversion optional, only needed by folks who haven't already checked and case the raw string.

👍 

> but I may propose a name change in the future :)

That's okay. afaic there can be many names, and it's unlikely everyone will be happy with any one name.
https://en.wikipedia.org/wiki/Law_of_triviality